### PR TITLE
bpo-42262: Add Py_NewRef() function

### PR DESCRIPTION
**Pull request by vstinner**
Add Py_NewRef() function to get a strong reference to an object.

https://bugs.python.org/issue42262

### CODE REVIEW

**Comment by vstinner** [comment]
Ubuntu failed with a download error:

E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/b/babeltrace/libbabeltrace1_1.5.5-1_amd64.deb  Could not connect to azure.archive.ubuntu.com:80 (52.252.75.106), connection timed out

**Comment by vstinner** [comment]
I closed/reopened the PR to re-run the Ubuntu job.

**Comment by vstinner** [comment]
cc @methane @serhiy-storchaka

**Review by methane — COMMENTED** [review]
Would you improve the doc to describe the difference to Py_INCREF?

**Comment by vstinner** [comment]
@methane:
> Would you improve the doc to describe the difference to Py_INCREF?

I rewrote the function documentation. Is the difference between Py_INCREF() and Py_NewRef() more explicit now?

I also added Py_XNewRef() which accepts NULL.

File: Objects/object.c [hunk 1]
```diff
@@ -136,6 +136,21 @@ Py_DecRef(PyObject *o)
     Py_XDECREF(o);
 }
 
+#undef Py_NewRef
+#undef Py_XNewRef
```
Comment (Jongy) [inline Objects/object.c]: Perhaps redefine these 2 macros after defining the 2 functions, so this file can enjoy the `static inline` version as well?

**Comment by methane** [comment]
> I rewrote the function documentation. Is the difference between Py_INCREF() and Py_NewRef() more explicit now?

LGTM.

**Review by methane — APPROVED** [review]

File: Objects/object.c (hunk 1 above)
Comment (vstinner) [reply Objects/object.c]: Good idea. I moved the definitions at the bottom of object.c, so object.c gets the fast static inlined flavor as well.

Comment (Jongy) [reply Objects/object.c]: > I moved the definitions at the bottom of object.c, 

Even better :)

**Comment by vstinner** [comment]
Thanks for the review @methane and @Jongy !

**Pull request merged**